std-bits/base/src/main/java/org/enso/base/cache/LRUCache.java

```diff
@@ -0,0 +1,311 @@
+    for (var mapEntry : sortedEntries) {
+      if (totalSize <= maxTotalCacheSize) {
+        break;
+      }
+      toRemove.add(mapEntry);
+      totalSize -= mapEntry.getValue().size();
+    }
+    assert totalSize <= maxTotalCacheSize;
+    removeCacheEntries(toRemove);
+  }
+
+  private SortedSet<Map.Entry<String, CacheEntry<M>>> getSortedEntries() {
+    var sortedEntries = new TreeSet<Map.Entry<String, CacheEntry<M>>>(cacheEntryLRUComparator);
+    sortedEntries.addAll(cache.entrySet());
+    return sortedEntries;
+  }
+
+  /** Remove least-recently used entries until the total cache size is under the limit. */
+  private void removeFilesToSatisfyLimit() {
+    makeRoomFor(0L);
+  }
+
+  private long getTotalCacheSize() {
+    return cache.values().stream().collect(Collectors.summingLong(e -> e.size()));
+  }
+
+  private long getMaxFileSize() {
+    return maxFileSizeOverrideTestOnly.orElse(maxFileSize);
+  }
+
+  private long getMaxTotalCacheSize() {
+    return maxTotalCacheSizeOverrideTestOnly.orElse(maxTotalCacheSize);
+  }
+
+  public int getNumEntries() {
+    return cache.size();
+  }
+
+  public List<Long> getFileSizesTestOnly() {
+    return new ArrayList<>(
+        cache.values().stream().map(CacheEntry::size).collect(Collectors.toList()));
+  }
+
+  public void setNowOverrideTestOnly(ZonedDateTime nowOverride) {
+    nowOverrideTestOnly = Optional.of(nowOverride);
+  }
+
+  public void clearNowOverrideTestOnly() {
+    nowOverrideTestOnly = Optional.empty();
+  }
+
+  public void setMaxFileSizeOverrideTestOnly(long maxFileSizeOverrideTestOnly_) {
+    assert maxFileSizeOverrideTestOnly_ <= maxFileSize;
+    maxFileSizeOverrideTestOnly = Optional.of(maxFileSizeOverrideTestOnly_);
+  }
+
+  public void clearMaxFileSizeOverrideTestOnly() {
+    maxFileSizeOverrideTestOnly = Optional.empty();
+  }
+
+  public void setMaxTotalCacheSizeOverrideTestOnly(long maxTotalCacheSizeOverrideTestOnly_) {
+    assert maxTotalCacheSizeOverrideTestOnly_ <= maxTotalCacheSize;
+    maxTotalCacheSizeOverrideTestOnly = Optional.of(maxTotalCacheSizeOverrideTestOnly_);
+  }
+
+  public void clearMaxTotalCacheSizeOverrideTestOnly() {
+    maxTotalCacheSizeOverrideTestOnly = Optional.empty();
+  }
+
+  private ZonedDateTime getNow() {
+    return nowOverrideTestOnly.orElse(ZonedDateTime.now());
+  }
+
+  private record CacheEntry<M>(File responseData, M metadata, long size, ZonedDateTime expiry) {}
+
+  public record Item<M>(
+      InputStream stream, M metadata, Optional<Long> sizeMaybe, Optional<Integer> ttl) {
+
+    public boolean shouldCache() {
+      return ttl.isPresent();
+    }
+  }
+
+  public record CacheResult<M>(InputStream inputStream, M metadata) {}
+
+  public interface ItemBuilder<M> {
+    String makeCacheKey();
+
+    /** Returning an Item with no TTL indicates that the data should not be cached. */
+    Item<M> buildItem() throws IOException, InterruptedException;
+  }
```

Comment:
I'd appreciate some docs on this class, or some clearer names (but I don't have too good ideas, so docs is always the fallback). 

Currently, it is not clear to me from the method names that `buildItem` is called only if no cached element was found.